helm: opentelemetry addon allow configuration of registry with setting tag

patst opened the pull request:
<!--- Provide a general summary of your changes in the Title above --->
<!--- Please don't @-mention people in PR or commit messages (do so in an additional comment). --->

## What this PR does / why we need it:
<!--- Why is this change required? What problem does it solve? -->
I am trying to get the OpenTelemetry integration running and I am using the helm chart.

I cannot use public container registries directly, but have a cache container registry in between.

Having the definition of an image like this:
```yaml
controller:
    opentelemetry:
        image: registry.k8s.io/ingress-nginx/opentelemetry:v20230312-helm-chart-4.5.2-28-g66a760794@sha256:40f766ac4a9832f36f217bb0e98d44c8d38faeccbfe861fbc1a76af7e9ab257f
```
(see  https://github.com/kubernetes/ingress-nginx/blob/e4a66fd2f625de3bcc7aaa793b31f529a0662009/charts/ingress-nginx/values.yaml#LL541C9-L541C9 ) is inconvenient for me, since I can only overwrite the complete key and have to worry about updating the tag and sha values from there on.

The other places in the helm chart reference images like this:
```yaml
controller:
    opentelemetry:
        image:
            registry: registry.k8s.io # can but my private registry here
            image: ingress-nginx/opentelemetry
            tag: "v20230312-helm-chart-4.5.2-28-g66a760794"
            digest: sha256:40f766ac4a9832f36f217bb0e98d44c8d38faeccbfe861fbc1a76af7e9ab257f
```
This allows me just to change the registry value to point to my private one.


<!--- If it fixes an open issue, please link to the issue here. -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] CVE Report (Scanner found CVE and adding report)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation only

## Which issue/s this PR fixes
<!--
(optional, in `fixes #<issue number>` format, will close that issue when PR gets merged):

fixes #
-->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Tested locally with `helm template ...` and the templated value for the image looked the same after my change

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x ] My change requires a change to the documentation.
- [x ] I have updated the documentation accordingly.
- [x ] I've read the [CONTRIBUTION](https://github.com/kubernetes/ingress-nginx/blob/main/CONTRIBUTING.md) guide
- [ ] I have added unit and/or e2e tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] Added Release Notes.

## Does my pull request need a release note?
Any user-visible or operator-visible change qualifies for a release note. This could be a:

- CLI change
- API change
- UI change
- configuration schema change
- behavioral change
- change in non-functional attributes such as efficiency or availability, availability of a new platform
- a warning about a deprecation
- fix of a previous Known Issue
- fix of a vulnerability (CVE)

No release notes are required for changes to the following:

- Tests
- Build infrastructure
- Fixes for unreleased bugs

For more tips on writing good release notes, check out the [Release Notes Handbook](https://github.com/kubernetes/sig-release/tree/master/release-team/role-handbooks/release-notes)

<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

longwuyuan commented:
Hi @patst,
Thanks for contribution.
Can you copy paste the complete data for a scene where this change was used.

Like show all the commands and outputs ranging from showing a modified
values yaml, helm install using the modified yaml, controller installed
successfully with otel, otlel container running, otel feature running etc
etc. Including logs.

That could a reader and approvers to see a successful install and
functioning first, before they test it themself.

But personally I feel you can not escape an edit in any case. So which line
you edit is not really a huge improvement for the project.

On Wed, 22 Mar, 2023, 1:16 am Kubernetes Prow Robot, <
***@***.***> wrote:

> Hi @patst <https://github.com/patst>. Thanks for your PR.
>
> I'm waiting for a kubernetes <https://github.com/orgs/kubernetes/people>
> member to verify that this patch is reasonable to test. If it is, they
> should reply with /ok-to-test on its own line. Until that is done, I will
> not automatically test new commits in this PR, but the usual testing
> commands by org members will still work. Regular contributors should join
> the org <https://git.k8s.io/community/community-membership.md#member> to
> skip this step.
>
> Once the patch is verified, the new status will be reflected by the
> ok-to-test label.
>
> I understand the commands that are listed here
> <https://go.k8s.io/bot-commands?repo=kubernetes%2Fingress-nginx>.
>
> Instructions for interacting with me using PR comments are available here
> <https://git.k8s.io/community/contributors/guide/pull-requests.md>. If
> you have questions or suggestions related to my behavior, please file an
> issue against the kubernetes/test-infra
> <https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:>
> repository.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/ingress-nginx/pull/9773#issuecomment-1478487351>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABGZVWRUXNKMAFMWDMDVSXLW5IARBANCNFSM6AAAAAAWC47ZAM>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

strongjz commented:
You need to update the deployment and the daemonset image for otel. https://github.com/kubernetes/ingress-nginx/blob/main/charts/ingress-nginx/templates/controller-deployment.yaml#L193

strongjz commented:
@patst can you rebase?

patst commented:
@strongjz  I rebased the change

@longwuyuan  thanks for your feedback. I added a test YAML file in the `ci` folder. The value seems minor, but the image definition is more aligned to the other definitions (see https://github.com/kubernetes/ingress-nginx/blob/e09e40af1ad33cf748ee92502f61ec2b6f741b96/charts/ingress-nginx/values.yaml#L18-L27 or https://github.com/kubernetes/ingress-nginx/blob/e09e40af1ad33cf748ee92502f61ec2b6f741b96/charts/ingress-nginx/values.yaml#L595-L603 ). 

Having the registry, tag and sha value in one attribute is very annoying, because usually we only want to override the repository url.

Without this change we are forced to check for each helm chart update if the tag/sha value did change as well. We could avoid that, if we are able to just replace the repository name. 

Installation output: (`helm install ps-oteltest ./charts/ingress-nginx -f ./charts/ingress-nginx/ci/deployment-opentelemetry-customregistry-values.yaml `)
```
NAME: ps-oteltest
LAST DEPLOYED: Tue Apr 18 09:24:46 2023
NAMESPACE: default
STATUS: deployed
REVISION: 1
TEST SUITE: None
NOTES:
The ingress-nginx controller has been installed.
It may take a few minutes for the LoadBalancer IP to be available.
You can watch the status by running 'kubectl --namespace default get services -o wide -w ps-oteltest-ingress-nginx-controller'

An example Ingress that makes use of the controller:
  apiVersion: networking.k8s.io/v1
  kind: Ingress
  metadata:
    name: example
    namespace: foo
  spec:
    ingressClassName: nginx
    rules:
      - host: www.example.com
        http:
          paths:
            - pathType: Prefix
              backend:
                service:
                  name: exampleService
                  port:
                    number: 80
              path: /
    # This section is only required if TLS is to be enabled for the Ingress
    tls:
      - hosts:
        - www.example.com
        secretName: example-tls

If TLS is enabled for the Ingress, a Secret containing the certificate and key must also be provided:

  apiVersion: v1
  kind: Secret
  metadata:
    name: example-tls
    namespace: foo
  data:
    tls.crt: <base64 encoded cert>
    tls.key: <base64 encoded key>
  type: kubernetes.io/tls

```

Pod description after installation: (`kubectl describe pod ps-oteltest-ingress-nginx-controller-5c9dffd6d5-29vdz  `)
```        
Name:             ps-oteltest-ingress-nginx-controller-5c9dffd6d5-29vdz
Namespace:        default
Priority:         0
Service Account:  ps-oteltest-ingress-nginx
Node:             colima/192.168.5.15
Start Time:       Tue, 18 Apr 2023 09:24:51 +0200
Labels:           app.kubernetes.io/component=controller
                  app.kubernetes.io/instance=ps-oteltest
                  app.kubernetes.io/managed-by=Helm
                  app.kubernetes.io/name=ingress-nginx
                  app.kubernetes.io/part-of=ingress-nginx
                  app.kubernetes.io/version=1.7.0
                  helm.sh/chart=ingress-nginx-4.6.0
                  pod-template-hash=5c9dffd6d5
Annotations:      <none>
Status:           Running
IP:               172.17.0.6
IPs:
  IP:           172.17.0.6
Controlled By:  ReplicaSet/ps-oteltest-ingress-nginx-controller-5c9dffd6d5
Init Containers:
  opentelemetry:
    Container ID:  docker://eee15bc4f7135f1c5e9a693b275aca63d554bf95b9d8e108ed65788af3bd0dd0
    Image:         registry.k8s.io/ingress-nginx/opentelemetry:v20230312-helm-chart-4.5.2-28-g66a760794@sha256:40f766ac4a9832f36f217bb0e98d44c8d38faeccbfe861fbc1a76af7e9ab257f
    Image ID:      docker-pullable://registry.k8s.io/ingress-nginx/opentelemetry@sha256:40f766ac4a9832f36f217bb0e98d44c8d38faeccbfe861fbc1a76af7e9ab257f
    Port:          <none>
    Host Port:     <none>
    Command:
      sh
      -c
      /usr/local/bin/init_module.sh
    State:          Terminated
      Reason:       Completed
      Exit Code:    0
      Started:      Tue, 18 Apr 2023 09:24:52 +0200
      Finished:     Tue, 18 Apr 2023 09:24:52 +0200
    Ready:          True
    Restart Count:  0
    Environment:    <none>
    Mounts:
      /modules_mount from modules (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-nfd5x (ro)
Containers:
  controller:
    Container ID:  docker://7a8ade790020de6f337e3dff379e694b9ef9bcc62b80a6f82275070bed5687f9
    Image:         registry.k8s.io/ingress-nginx/controller:v1.7.0@sha256:7612338342a1e7b8090bef78f2a04fffcadd548ccaabe8a47bf7758ff549a5f7
    Image ID:      docker-pullable://registry.k8s.io/ingress-nginx/controller@sha256:7612338342a1e7b8090bef78f2a04fffcadd548ccaabe8a47bf7758ff549a5f7
    Ports:         80/TCP, 443/TCP, 8443/TCP
    Host Ports:    0/TCP, 0/TCP, 0/TCP
    Args:
      /nginx-ingress-controller
      --publish-service=$(POD_NAMESPACE)/ps-oteltest-ingress-nginx-controller
      --election-id=ps-oteltest-ingress-nginx-leader
      --controller-class=k8s.io/ingress-nginx
      --ingress-class=nginx
      --configmap=$(POD_NAMESPACE)/ps-oteltest-ingress-nginx-controller
      --validating-webhook=:8443
      --validating-webhook-certificate=/usr/local/certificates/cert
      --validating-webhook-key=/usr/local/certificates/key
    State:          Running
      Started:      Tue, 18 Apr 2023 09:24:53 +0200
    Ready:          True
    Restart Count:  0
    Requests:
      cpu:      100m
      memory:   90Mi
    Liveness:   http-get http://:10254/healthz delay=10s timeout=1s period=10s #success=1 #failure=5
    Readiness:  http-get http://:10254/healthz delay=10s timeout=1s period=10s #success=1 #failure=3
    Environment:
      POD_NAME:       ps-oteltest-ingress-nginx-controller-5c9dffd6d5-29vdz (v1:metadata.name)
      POD_NAMESPACE:  default (v1:metadata.namespace)
      LD_PRELOAD:     /usr/local/lib/libmimalloc.so
    Mounts:
      /modules_mount from modules (rw)
      /usr/local/certificates/ from webhook-cert (ro)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-nfd5x (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             True 
  ContainersReady   True 
  PodScheduled      True 
Volumes:
  modules:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:     
    SizeLimit:  <unset>
  webhook-cert:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  ps-oteltest-ingress-nginx-admission
    Optional:    false
  kube-api-access-nfd5x:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   Burstable
Node-Selectors:              kubernetes.io/os=linux
Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type    Reason     Age   From                      Message
  ----    ------     ----  ----                      -------
  Normal  Scheduled  41s   default-scheduler         Successfully assigned default/ps-oteltest-ingress-nginx-controller-5c9dffd6d5-29vdz to colima
  Normal  Pulled     40s   kubelet                   Container image "registry.k8s.io/ingress-nginx/opentelemetry:v20230312-helm-chart-4.5.2-28-g66a760794@sha256:40f766ac4a9832f36f217bb0e98d44c8d38faeccbfe861fbc1a76af7e9ab257f" already present on machine
  Normal  Created    40s   kubelet                   Created container opentelemetry
  Normal  Started    40s   kubelet                   Started container opentelemetry
  Normal  Pulled     39s   kubelet                   Container image "registry.k8s.io/ingress-nginx/controller:v1.7.0@sha256:7612338342a1e7b8090bef78f2a04fffcadd548ccaabe8a47bf7758ff549a5f7" already present on machine
  Normal  Created    39s   kubelet                   Created container controller
  Normal  Started    39s   kubelet                   Started container controller
  Normal  RELOAD     37s   nginx-ingress-controller  NGINX reload triggered due to a change in configuration
```

strongjz commented on charts/ingress-nginx/ci/deployment-opentelemetry-customregistry-values.yaml:
```diff
@@ -0,0 +1,6 @@
+  opentelemetry:
+    enabled: true
+    image:
+      registry: mycontainerreg.some.io
```
https://github.com/kubernetes/ingress-nginx/actions/runs/4730554702/jobs/8394620062?pr=9773#step:5:135

patst replied:
I set the registry + image (tag and sha will be used from the values.yaml)

patst commented:
The CI test failed, I see no direct error in the logs at https://github.com/kubernetes/ingress-nginx/actions/runs/4731681976/jobs/8397078869?pr=9773 . Maybe it is a technical error during pipeline execution?

strongjz commented:
@patst can you rebase? The helm failures should be resolved now, there was an issue with cert-manager deployment. 

-James

patst commented:
> @patst can you rebase? The helm failures should be resolved now, there was an issue with cert-manager deployment.
> 
> -James

@strongjz :yep, sure thanks for taking a look at it. I did the rebase, but the tests seem to fail anyway.

```
git fetch origin
git merge origin/Main
=> Already up to date.
```

I see this in the logs;: (https://github.com/kubernetes/ingress-nginx/actions/runs/4823089524/jobs/8591289654?pr=9773#step:11:303 )
```
29s         Warning   ErrInitIssuer        issuer/ingress-nginx-c6xql0f8oh-root-issuer                                                 cert-manager-issuers                                 Error initializing issuer: secret "ingress-nginx-c6xql0f8oh-root-cert" not found  
```

Any idea what I have missed?

strongjz commented:
@patst we have fixed the flaky test by replacing httpbin with httpbun in #9919 please rebase to get those changes.

strongjz commented:
/triage accepted
/priority backlog

Gacko commented:
@patst Can you please rebase this PR? There have been some changes especially to the OpenTelemetry and the `extraModules` part and I was also already thinking about having images split, so I'd really like to collaborate on this!

/assign Gacko

Gacko commented on charts/ingress-nginx/values.yaml:
```diff
@@ -546,7 +546,11 @@ controller:
   opentelemetry:
     enabled: false
-    image: registry.k8s.io/ingress-nginx/opentelemetry:v20230312-helm-chart-4.5.2-28-g66a760794@sha256:40f766ac4a9832f36f217bb0e98d44c8d38faeccbfe861fbc1a76af7e9ab257f
+    image:
+      registry: registry.k8s.io
+      image: ingress-nginx/opentelemetry
+      tag: "v20230312-helm-chart-4.5.2-28-g66a760794"
+      digest: sha256:40f766ac4a9832f36f217bb0e98d44c8d38faeccbfe861fbc1a76af7e9ab257f
```
Basically this is a breaking change, but it makes sense. Could you please also apply this to the `extraModules` array and do the whole concatenation in the `extraModules` template in the `_helpers.tpl`?

patst replied:
thanks for the feedback!

I moved the concatenation to the `extraModules` template. I reused the existing template function for the digest.

I adapted the `extraModules` example in the values yaml to match the new pattern

Gacko commented on charts/ingress-nginx/templates/controller-deployment.yaml:
```diff
@@ -189,8 +189,9 @@ spec:
       {{- if .Values.controller.opentelemetry.enabled}}
-          {{ $otelContainerSecurityContext := $.Values.controller.opentelemetry.containerSecurityContext | default $.Values.controller.containerSecurityContext }}
-          {{- include "extraModules" (dict "name" "opentelemetry" "image" .Values.controller.opentelemetry.image "containerSecurityContext" $otelContainerSecurityContext) | nindent 8}}
+          {{- $otelContainerSecurityContext := $.Values.controller.opentelemetry.containerSecurityContext | default $.Values.controller.containerSecurityContext }}
+          {{- $fullImage := printf "%s/%s:%s@%s" .Values.controller.opentelemetry.image.registry .Values.controller.opentelemetry.image.image .Values.controller.opentelemetry.image.tag .Values.controller.opentelemetry.image.digest }}
+          {{- include "extraModules" (dict "name" "opentelemetry" "image" $fullImage "containerSecurityContext" $otelContainerSecurityContext) | nindent 8}}
```
See https://github.com/kubernetes/ingress-nginx/pull/9773/files#diff-46ba6538a37f39ee40cb77261314c803fc91d830c8154a9f08d0f6b2bfe31f75L549-R553.

Gacko replied:
Also this part needs rebasing.

patst replied:
rebased and used the templating for the image from inside the `extraModules` template

Gacko commented on charts/ingress-nginx/templates/controller-daemonset.yaml:
```diff
@@ -184,6 +184,7 @@ spec:
       {{- end }}
       {{- if .Values.controller.opentelemetry.enabled}}
           {{ $otelContainerSecurityContext := $.Values.controller.opentelemetry.containerSecurityContext | default $.Values.controller.containerSecurityContext }}
+          {{ $fullImage := printf "%s/%s:%s@%s" .Values.controller.opentelemetry.image.registry .Values.controller.opentelemetry.image.image .Values.controller.opentelemetry.image.tag .Values.controller.opentelemetry.image.digest }}
```
Same as for https://github.com/kubernetes/ingress-nginx/pull/9773/files#diff-147eaabb2f1187e85b285ac15fc2542b761ce6b8b0ed100102df85b878701575L192-R194.

Gacko commented on charts/ingress-nginx/README.md:
```diff
@@ -371,7 +371,10 @@ As of version `1.26.0` of this chart, by simply not providing any clusterIP valu
 | controller.opentelemetry.containerSecurityContext.allowPrivilegeEscalation | bool | `false` |  |
 | controller.opentelemetry.enabled | bool | `false` |  |
-| controller.opentelemetry.image | string | `"registry.k8s.io/ingress-nginx/opentelemetry:v20230312-helm-chart-4.5.2-28-g66a760794@sha256:40f766ac4a9832f36f217bb0e98d44c8d38faeccbfe861fbc1a76af7e9ab257f"` |  |
+| controller.opentelemetry.image.digest | string | `"sha256:40f766ac4a9832f36f217bb0e98d44c8d38faeccbfe861fbc1a76af7e9ab257f"` |  |
+| controller.opentelemetry.image.image | string | `"ingress-nginx/opentelemetry"` |  |
+| controller.opentelemetry.image.registry | string | `"registry.k8s.io"` |  |
+| controller.opentelemetry.image.tag | string | `"v20230312-helm-chart-4.5.2-28-g66a760794"` |  |
```
Please use the latest tag here.

patst commented on charts/ingress-nginx/values.yaml:
```diff
@@ -607,11 +610,13 @@ controller:
-      digest: sha256:40f766ac4a9832f36f217bb0e98d44c8d38faeccbfe861fbc1a76af7e9ab257f
+      tag: "v20230721-3e2062ee5"
+      digest: sha256:13bee3f5223883d3ca62fee7309ad02d22ec00ff0d7033e3e9aca7a9f60fd472
+    distroless: true
```
fair point!
Moved it under the `image` object as well and updated the helm-docs

patst commented:
@Gacko  resolved all the discussions and templating looks fine.

The integration-tests fail and I see the message `Error: Error installing charts: Error processing charts`, but cannot find a log message giving me more details.

Do you see something?

Gacko commented:
Not sure, but it could be related to the values in the [CI directory](https://github.com/kubernetes/ingress-nginx/tree/main/charts/ingress-nginx/ci). Some of them contain `extraModules` and since you changed the image format, it could be their image config is not working anymore.

patst commented:
> Not sure, but it could be related to the values in the [CI directory](https://github.com/kubernetes/ingress-nginx/tree/main/charts/ingress-nginx/ci). Some of them contain `extraModules` and since you changed the image format, it could be their image config is not working anymore.

thanks for the hint! That was correct and I fixed the test cases and reworked the templating a bit to make the repository and tag optional.

Nevertheless, one testcase is failing and taking a look at the logs I guess it is related to the one i added `charts/ingress-nginx/ci/deployment-opentelemetry-customregistry-values.yaml`.
The templating does work fine locally and I see no specific entry in the errors log which help me debugging the issues.

Any ideas how to tackle that problem @Gacko ?

Gacko commented:
Regarding the image templating:

```
{{/*
Extra modules.
*/}}
{{- define "extraModules" -}}
- name: {{ .name }}
  {{- with .image }}
  image: {{ if .repository }}{{ .repository }}{{ else }}{{ .registry }}/{{ .image }}{{ end }}:{{ .tag }}{{ if .digest }}@{{ .digest }}{{ end }}
  command:
  {{- if .distroless }}
    - /init_module
  {{- else }}
    - sh
    - -c
    - /usr/local/bin/init_module.sh
  {{- end }}
  {{- end }}
  {{- if .containerSecurityContext }}
  securityContext: {{ toYaml .containerSecurityContext | nindent 4 }}
  {{- end }}
  {{- if .resources }}
  resources: {{ toYaml .resources | nindent 4 }}
  {{- end }}
  volumeMounts:
    - name: modules
      mountPath: /modules_mount
{{- end -}}
```

This is in alignment with how images are being templated in the controller's `Deployment`/`DaemonSet`, default backend's `Deployment` and webhook patch `Job`s. I'll file an issue and a PR about the minor formatting differences soon.

Making the tag optional adds the possibility of not defining one at all, which leads to using `latest` and image pull policy `Always`: https://kubernetes.io/docs/concepts/containers/images/#image-pull-policy

This is not recommended, especially for production use cases, as you never know which version really is in use then. Additionally it can slow down or block pod creation since `kubelet` first needs to check the digest every time it starts a container.

Gacko commented:
Making registry & tag mandatory would also require changes in the CI values. Hm...

Apart from that I feel like webhook secret patch job is not completing. Gonna have a look. The only real difference I spot is you're not using `controller.service.type: ClusterIP`. But this should not prevent the job from completing. Also I'm not sure if this is really the root cause.

Gacko commented on charts/ingress-nginx/values.yaml:
```diff
@@ -587,8 +587,11 @@ controller:
   # - name: mytestmodule
-  #   image: registry.k8s.io/ingress-nginx/mytestmodule
-  #   distroless: false
+  #   image:
+  #     registry: registry.k8s.io
+  #     image: ingress-nginx/mytestmodule
+  #     tag: latest
+  #     distroless: false
```
```suggestion
  #   image:
  #     registry: registry.k8s.io
  #     image: ingress-nginx/mytestmodule
  #     ## for backwards compatibility consider setting the full image url via the repository value below
  #     ## use *either* current default registry/image or repository format or installing chart by providing the values.yaml will fail
  #     ## repository:
  #     tag: "v1.0.0"
  #     digest: ""
  #     distroless: false
```

Gacko commented on charts/ingress-nginx/values.yaml:
```diff
@@ -608,8 +611,12 @@ controller:
     name: opentelemetry
-    image: registry.k8s.io/ingress-nginx/opentelemetry:v20230721-3e2062ee5@sha256:13bee3f5223883d3ca62fee7309ad02d22ec00ff0d7033e3e9aca7a9f60fd472
-    distroless: true
+    image:
+      registry: registry.k8s.io
+      image: ingress-nginx/opentelemetry
+      tag: "v20230721-3e2062ee5"
+      digest: sha256:13bee3f5223883d3ca62fee7309ad02d22ec00ff0d7033e3e9aca7a9f60fd472
+      distroless: true
```
```suggestion
    image:
      registry: registry.k8s.io
      image: ingress-nginx/opentelemetry
      ## for backwards compatibility consider setting the full image url via the repository value below
      ## use *either* current default registry/image or repository format or installing chart by providing the values.yaml will fail
      ## repository:
      tag: "v20230721-3e2062ee5"
      digest: sha256:13bee3f5223883d3ca62fee7309ad02d22ec00ff0d7033e3e9aca7a9f60fd472
      distroless: true
```

Gacko commented on charts/ingress-nginx/ci/daemonset-extra-modules.yaml:
```diff
@@ -7,4 +7,5 @@ controller:
   extraModules:
     - name: opentelemetry
-      image: busybox
+      image:
+        image: busybox
```
```suggestion
      image:
        registry: registry.k8s.io
        image: busybox
        tag: latest
```

Gacko replied:
Actually I'm wondering how does can come up as the default `busybox` image from Docker Hub also does not contain the `/usr/local/bin/init_module.sh`.

Gacko commented on charts/ingress-nginx/ci/deployment-extra-modules-default-container-sec-context.yaml:
```diff
@@ -9,4 +9,5 @@ controller:
   extraModules:
     - name: opentelemetry
-      image: busybox
+      image:
+        image: busybox
```
```suggestion
      image:
        registry: registry.k8s.io
        image: busybox
        tag: latest
```

Gacko commented on charts/ingress-nginx/ci/deployment-extra-modules-specific-container-sec-context.yaml:
```diff
@@ -7,6 +7,7 @@ controller:
   extraModules:
     - name: opentelemetry
-      image: busybox
+      image:
+        image: busybox
```
```suggestion
      image:
        registry: registry.k8s.io
        image: busybox
        tag: latest
```

Gacko commented on charts/ingress-nginx/ci/deployment-extra-modules.yaml:
```diff
@@ -7,4 +7,5 @@ controller:
   extraModules:
     - name: opentelemetry
-      image: busybox
+      image:
+        image: busybox
```
```suggestion
      image:
        registry: registry.k8s.io
        image: busybox
        tag: latest
```

Gacko reviewed (commented):
Left some suggestions, hope I'm not overwhelming you.

Gacko commented:
Thank you for updating everything! It looks awesome now and I really hope the tests are passing. Is there anything left you'd like to change in this PR or is it safe to merge after tests pass?

Gacko commented:
Here we go! Thanks for your contribution!
